### PR TITLE
prewrite only when keys are not exist

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
Signed-off-by: zhangjinpeng1987 <zhangjinpeng@pingcap.com>

## What have you changed? (mandatory)

Add support for Insert operation, prewrite will success only when all keys are not exist. With this feature, TiDB doesn't have to use a get/batch get to check if keys are exist when handling insert statement.

## What are the type of the changes? (mandatory)

- New feature

## Does this PR affect documentation (docs) update? (mandatory)

No

## Does this PR affect tidb-ansible update? (mandatory)

No

## Refer to a related PR or issue link (optional)
https://github.com/pingcap/kvproto/pull/343 Should be merged first.

## Performance test
Sysbench 32 tables, each table has 1000,000 rows, 32 threads insert
cpu-usage
grpc-count

### CODE REVIEW

**Comment by siddontang** [comment]
any performance improved?

**Comment by zhangjinpeng1987** [comment]
> any performance improved?

@siddontang Please see the performance test result.

**Comment by zhangjinpeng1987** [comment]
> Seems that it is not covered by any test

@Connor1996 Done

File: src/storage/mod.rs [hunk 1]
```diff
@@ -80,7 +80,7 @@ pub fn is_short_value(value: &[u8]) -> bool {
 
 #[derive(Debug, Clone)]
 pub enum Mutation {
-    Put((Key, Value)),
+    Put((Key, Value, bool /*should not exist*/)),
```
Comment (BusyJay) [inline src/storage/mod.rs]: `()` is unnecessary. And struct is preferred in this case.

**Comment by ngaut** [comment]
PTAL @disksing @coocood @AndreMouche

**Comment by coocood** [comment]
@zhangjinpeng1987 
Why not adding another mutation type like Mutation::'Insert' instead of adding a boolean field?

**Comment by zhangjinpeng1987** [comment]
> Why not adding another mutation type like Mutation::Insert instead of adding a boolean field?

After discussing with @tiancaiamao and @coocood , we add an `Insert` operation to keep `Put` operation pure. https://github.com/pingcap/kvproto/pull/353

**Comment by zhangjinpeng1987** [comment]
@coocood @disksing @tiancaiamao PTAL

**Comment by coocood** [comment]
This method need to change too?
https://github.com/tikv/tikv/blob/master/src/server/service/kv.rs#L1660

**Comment by coocood** [comment]
Never mind, the `Insert` is converted to `Put` when converting to LockType.

File: src/storage/mvcc/txn.rs [hunk 2]
```diff
@@ -139,16 +139,35 @@ impl<S: Snapshot> MvccTxn<S> {
         self.writes.push(Modify::Delete(CF_WRITE, key));
     }
 
+    fn key_exist(&mut self, key: &Key, ts: u64) -> Result<bool> {
+        Ok(self.reader.skip_lock_get(&key, ts)?.is_some())
```
Comment (coocood) [inline src/storage/mvcc/txn.rs]: The `skip_lock_get` reads default CF, we don't need to do that.

Comment (coocood) [reply src/storage/mvcc/txn.rs]: I mean make `skip_lock_get` skips Rollback and Lock, return `Write`.

File: src/storage/mvcc/reader/mod.rs [hunk 3]
```diff
@@ -234,10 +234,6 @@ impl<S: Snapshot> MvccReader<S> {
             IsolationLevel::SI => ts = self.check_lock(key, ts)?,
             IsolationLevel::RC => {}
         }
-        self.skip_lock_get(key, ts)
-    }
-
-    pub fn skip_lock_get(&mut self, key: &Key, mut ts: u64) -> Result<Option<Value>> {
         loop {
```
Comment (coocood) [inline src/storage/mvcc/reader/mod.rs]: We can call `get_write` here.

**Comment by coocood** [comment]
LGTM

**Comment by tiancaiamao** [comment]
Where is the update? @zhangjinpeng1987

**Comment by zhangjinpeng1987** [comment]
> Where is the `Update` command?

@tiancaiamao It should be in another PR.

**Comment by ngaut** [comment]
PTAL @disksing

**Comment by zhangjinpeng1987** [comment]
@disksing @MyonKeminta PTAL

File: src/storage/mvcc/reader/mod.rs [hunk 4]
```diff
@@ -812,4 +819,60 @@ mod tests {
...
+        let region = make_region(1, vec![], vec![]);
+        let db = open_db(path, true);
+        let mut engine = RegionEngine::new(Arc::clone(&db), region.clone());
+
+        let (k, v) = (b"k", b"v");
+        let m = Mutation::Put((Key::from_raw(k), v.to_vec()));
+        engine.prewrite(m, k, 1);
+        engine.commit(k, 1, 2);
+
+        engine.rollback(k, 5);
```
Comment (MyonKeminta) [inline src/storage/mvcc/reader/mod.rs]: Don't you need to prewrite something before calling rollback?

Comment (zhangjinpeng1987) [reply src/storage/mvcc/reader/mod.rs]: Yes

**Comment by zhangjinpeng1987** [comment]
@AndreMouche It will behavior the same as before.

**Comment by zhangjinpeng1987** [comment]
@disksing PTAL

File: src/storage/mvcc/txn.rs [hunk 5]
```diff
@@ -163,10 +174,18 @@ impl<S: Snapshot> MvccTxn<S> {
                             primary: primary.to_vec(),
                         });
                     }
+                    if should_not_exist {
+                        if write.write_type == WriteType::Put
+                            || (write.write_type != WriteType::Delete
+                                && self.key_exist(&key, write.start_ts - 1)?)
```
Comment (disksing) [inline src/storage/mvcc/txn.rs]: How about just `if should_not_exist && self.key_exist(&key, ts-1)`?

Comment (zhangjinpeng1987) [reply src/storage/mvcc/txn.rs]: We have got a latest write here, it may don't need to call `key_exist`.

**Pull request merged**